ARTESCA-3572 Allow the removal of Metalk8s archives in iso-manager

sayf-eddine-scality opened the pull request:
**Component**:

<!-- E.g. 'salt', 'containers', 'kubernetes', 'build', 'tests'... -->

**Context**: 
Allow the removal of the iso
**Summary**:
```
./iso-manager.sh --rm-archive /root/metalk8s-2.10.7.iso
```
**Acceptance criteria**: 


---

<!-- Declare one or more issues to close once this PR gets merged -->

Closes: #ISSUE_NUMBER

<!-- If you want to refer to an issue while not closing it, use:

See: #ISSUE_NUMBER

-->

sayf-eddine-scality committed:
scripts,salt: Allow ISO removal in iso-manager

This adds a `-r`/`--rm-archive` option to the script, and deprecates the
`--archive` in favor of `--add-archive` (for disambiguation).

It also slightly reworks our module extracting info from the
`product.txt` file in archives, to ensure the contents are valid at this
stage, instead of having error management in the
`metalk8s.archives.mounted` state.

bert-e commented:
# Hello sayf-eddine-scality,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @chengyanjin

* @lucieleonard

* @jbwatenbergscality

* @alexandre-allard-scality

* @gdemonet

* @teddyandrieux

* @nicolast

* @thomasdanan

TeddyAndrieux commented on scripts/iso-manager.sh:
```diff
@@ -84,6 +90,12 @@ _add_archives() {
 
+_remove_archives() {
+    for archive; do
+        $SALT_CALL metalk8s.configure_archive "$archive" remove=True
```
Not really due to this PR but it's really sad that even when you try to remove an archive, we need this one to be there and available 

IMO this line in the module shouldn't be there when we remove 
```python
    # raise if the archive does not exist or is invalid
    archive_info_from_product_txt(archive)
```

TeddyAndrieux commented on salt/_modules/metalk8s.py:
```diff
@@ -218,6 +218,34 @@ def archive_info_from_iso(path):
+    """List the already configured metalk8s archives"""
+    archives = {}
+    base_mount = "/srv/scality"
+    for path in os.listdir(base_mount):
```
Why do we check all directories, IMHO we should just check for mountpoint

TeddyAndrieux commented on salt/_modules/metalk8s.py:
```diff
@@ -218,6 +218,34 @@ def archive_info_from_iso(path):
+        archive = archive_info_from_product_txt(path)
+        # Check if the archive is really configured or just mounted
+        env_name = "metalk8s-{0}".format(archive["version"])
+        if not os.path.exists(f"/var/lib/metalk8s/repositories/conf.d/99-{env_name}-registry.inc"):
+            continue
```
I'm not sure we want to check this cause if for whatever reason this file does not exist we will not umount the ISO, to me we manage everything in `/srv/scality/` so if there is "something" mounted in this directory that we don't know we just umount it.

Especially since we first reconfigure the registry before actually umount the ISO it means we will never umount ISOs since when we run the state to umount the ISOs the registry file is already deleted

TeddyAndrieux commented on salt/metalk8s/archives/unmounted.sls:
```diff
@@ -0,0 +1,10 @@
+  mount.unmounted:
+    - name: {{ archive.path }}
+    - device: {{ archive.iso }}
+    - persist: true
```
It properly clean the fstab entry ? Cool :+1:

TeddyAndrieux replied:
Did you check this btw ?

gdemonet replied:
Yep, just checked (also looked at `/etc/fstab`, to be sure :innocent:):

```
          ID: Unmount '/srv/scality/releases/metalk8s-2.11.7.iso' from '/srv/scality/metalk8s-2.11.7'
    Function: mount.unmounted
        Name: /srv/scality/metalk8s-2.11.7
      Result: True
     Comment: Target was successfully unmounted. Removed target from fstab
     Started: 10:09:59.285132
    Duration: 262.827 ms
     Changes:   
              ----------
              persist:
                  purged
              umount:
                  True
```

TeddyAndrieux commented on salt/metalk8s/archives/unmounted.sls:
```diff
@@ -0,0 +1,10 @@
+{% set removable_archives = salt.metalk8s.get_removable_archives() %}
+{% for env, archive in removable_archives.items() %}
+{% if archive.iso %}
+Archive {{ archive.iso }} absent at {{ archive.path }}:
```
I know it comes from the state name from `mounted.sls` but I suggest putting a name a bit clearer
```suggestion
Unmount archive {{ archive.iso }} from {{ archive.path }}:
```

TeddyAndrieux commented on salt/metalk8s/archives/unmounted.sls:
```diff
@@ -0,0 +1,10 @@
+    - name: {{ archive.path }}
+    - device: {{ archive.iso }}
+    - persist: true
+{% endif %}
```
IMO we also want to remove the directory when we umount, only if the directory is empty

Something like that (not tested)
```yaml
Cleanup {{ archive.path }} directory:
  file.absent:
    - name: {{ archive.path }}
    - unless: file.path_exists_glob("{{ archive.path }}/*")
```

TeddyAndrieux commented on salt/metalk8s/repo/configured.sls:
```diff
@@ -48,3 +48,10 @@ Generate container registry configuration:
     - defaults:
         archives: {{ archives }}
+
+{% set removable_archives = salt.metalk8s.get_removable_archives() %}
+{% for env in removable_archives %}
+Remove config for env {{ env }}:
+    file.absent:
+      - name: {{ salt.file.join(repo.config.directory, '99-' ~ env ~ '-registry.inc') }}
+{% endfor %}
```
Again, to me we manage the full `repo.config.directory` so if we don't know a file in this directory we just remove it, since it's a bit complicated with solutions I suggest just checking all `99-metalk8s-*` we remove all files that are not expected

gdemonet replied:
:thinking: what if the Solution is called `metalk8s-solution-example`..? I'll go with a regex such as `99-metalk8s-[0-9].*`, should be enough.

Also, this state is a bit strange, we're not really managing all the includes, only adding one for the current version :confused:... I'll leave it like that but still confusing IMHO

gdemonet replied:
:thinking: what if the Solution is called `metalk8s-solution-example`..? I'll go with a regex such as `99-metalk8s-[0-9].*`, should be enough.

Also, this state is a bit strange, we're not really managing all the includes, only adding one for the current version :confused:... I'll leave it like that but still confusing IMHO

TeddyAndrieux replied:
Maybe solutions shouldn't have the "99-" prefix but another one, or yes just a simple regex

> Also, this state is a bit strange, we're not really managing all the includes, only adding one for the current version confused... I'll leave it like that but still confusing IMHO

That is my point, to me, we may want to change this state so that it basically "computes" the full directory content

TeddyAndrieux reviewed (commented):
Few comments (if you add salt module you also need to add unit tests :wink: )

bert-e commented:
# Branches have diverged



This pull request's source branch `improvement/ARTESCA-3572-handle-iso-removal` has diverged from
`development/123.0` by more than 50 commits.

To avoid any integration risks, please re-synchronize them using one of the
following solutions:

* Merge `origin/development/123.0` into `improvement/ARTESCA-3572-handle-iso-removal`
* Rebase `improvement/ARTESCA-3572-handle-iso-removal` onto `origin/development/123.0`

Note: If you choose to rebase, you may have to ask me to rebuild
integration branches using the `reset` command.

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @alexandre-allard-scality

* @jbwatenbergscality

* @nicolast

* @gdemonet

* @teddyandrieux

* @thomasdanan

* @lucieleonard

* @chengyanjin

gdemonet committed:
ci: Check ISO removal after upgrade

To make sure we can safely remove an old ISO, we do so in upgrade tests,
prior to running the test suite on the upgraded cluster.

TeddyAndrieux commented on CHANGELOG.md:
```diff
@@ -24,6 +28,13 @@
+
+- The long name `--archive` for the "add" option to the `iso-manager.sh` script
+  is deprecated in favor of `--add-archive` (shorthand `-a` is still valid).
+  This deprecated option will be removed in MetalK8s 123.1.0
```
To me, it should be removed in 124.0.0 cause you can upgrade from 2.11 to 123.1 (which means it gets removed directly with no "deprecations")

gdemonet replied:
Good point, I didn't remember we changed the "one minor at a time" to "one major at a time" :sweat_smile:

TeddyAndrieux commented on eve/main.yml:
```diff
@@ -2183,6 +2183,19 @@ stages:
+          command: >
+            ISO_PATH=$(
+            [[%(prop:nodes_count:-0)s -eq 0]] && echo "metalk8s.iso"
+            || echo "/archives/metalk8s.iso") &&
```
:cry: 
Can we get the path using salt instead ?
With `metalk8s.get_archives` module and `%(prop:product_promoted_version)s`
Something like that 
```console
$ salt-call slsutil.renderer --out=txt string="{{ salt.metalk8s.get_archives()['metalk8s-%(prop:product_promoted_version)s']['iso'] }}" | cut -c 8-
/archives/metalk8s.iso
```

gdemonet replied:
Will be simpler to maintain indeed, good idea :ok_hand:

TeddyAndrieux commented on salt/_modules/metalk8s.py:
```diff
@@ -218,6 +238,33 @@ def archive_info_from_iso(path):
+        except CommandExecutionError as exc:
+            log.info(
+                "Skipping %s (mounted at %s): not a valid MetalK8s archive"
+                "(failed to read/parse 'product.txt' contents): %s",
```
nitpick: Not sure it is worth mentioning "failed to read/parse" since we put the exception message, especially since you added a check on name and version (in those cases the read and parse of the product.txt worked well just not a valid archive)

gdemonet replied:
True, will change

TeddyAndrieux commented on salt/metalk8s/archives/mounted.sls:
```diff
@@ -1,27 +1,14 @@
-{%- for _, archive in salt.metalk8s.get_archives().items() %}
-  {%- if archive.iso %}
-    {%- if not archive.version %}
+{%- for archive in salt.metalk8s.get_archives().values() | selectattr("iso") %}
```
:+1:

TeddyAndrieux commented on salt/metalk8s/archives/unmounted.sls:
```diff
@@ -0,0 +1,31 @@
+Remove archive mountpoint '{{ archive.path }}':
+  file.absent:
+    - name: {{ archive.path }}
+    - unless:
+      - fun: file.path_exists_glob
+        path: {{ archive.path }}/*
```
So in this case we just silently ignore it and we do not remove the directory ? :thinking: 
Ok fine

gdemonet replied:
That was your suggestion, I just implemented it :innocent: But it makes sense to tell the user about it, I'll add a state for this.

TeddyAndrieux replied:
Oh, indeed :smile:

TeddyAndrieux commented on scripts/iso-manager.sh:
```diff
@@ -48,6 +57,13 @@ while (( "$#" )); do
 
+if [ -z ${ARCHIVE_DEPRECATED_OPTION} ]; then
+  echo "Warning: Usage of the '--archive' option is deprecated, and will be removed" \
+       "in the next MetalK8s minor version, 123.1.0. Please make sure to update your" \
```
ditto: 124.0.0

TeddyAndrieux commented on scripts/iso-manager.sh:
```diff
@@ -13,18 +13,27 @@ SALTENV=""
-    echo "-d/--dry-run:                    Run actions in dry run mode"
+    echo "-a/--add-archive <archive path>:     Path to an archive folder or ISO to add"
+    echo "    [DEPRECATED] You may also use --archive as an equivalent to the above."
+    echo "                 This option name will be removed in MetalK8s 123.1.0"
```
ditto: 124.0.0

TeddyAndrieux replied:
:eye: :smile:

gdemonet replied:
ooops

TeddyAndrieux commented on salt/metalk8s/archives/unmounted.sls:
```diff
@@ -0,0 +1,31 @@
+{# Remove mounted ISOs which were removed from configuration #}
```
I would have preferred a single state `available` (same as what we have for solutions) so that in the future we could have a single state to manage all archives (metalk8s + solutions)

But not really mandatory

gdemonet replied:
I thought about it, but right now we can't, because we want to first mount new ISOs, reconfigure the repo container, then unmount old ISOs (otherwise we may have issues with it, I think).

TeddyAndrieux replied:
Right make sense

TeddyAndrieux commented on salt/tests/unit/modules/files/test_metalk8s.yaml:
```diff
@@ -67,6 +67,45 @@ get_archives:
+  - mounts: {}
+    logs: []
+    result: {}
+  # 1. Nothing mounted under /srv/scality/metalk8s-*
```
Maybe just add another that looks like a solution that will be in `/srv/scality/`

I would have also added a "complex" test with at least 4 mount points 2 valid MetalK8s 1 invalid (skipped with logs) and one "other" mount just ignored because not under `/srv/scality/metalk8s-*`

TeddyAndrieux commented on scripts/iso-manager.sh:
```diff
@@ -84,6 +100,12 @@ _add_archives() {
 
+_remove_archives() {
+    for archive; do
+        $SALT_CALL metalk8s.configure_archive "$archive" remove=True
```
Not linked to this PR, but from what I see even if you use `--dry-run` we will still update the bootstrap config

gdemonet replied:
Added [`e2a95a2` (#3730)](https://github.com/scality/metalk8s/pull/3730/commits/e2a95a2aea8bb51a163e83bcf3a9b4c7ef6b4033) to handle this, but read the commit message, the behaviour will not be ideal.

TeddyAndrieux replied:
I'm not sure it's worth, I personally prefer the behavior we have today :confused:

gdemonet replied:
:cry:

gdemonet replied:
OK just removed it then, we'll have to think of a better way to manage this (or just close our eyes until we remove Salt :laughing:)

TeddyAndrieux replied:
![Alt Text](https://media.giphy.com/media/fvYR60mQCKt2x9hV3y/giphy.gif)

TeddyAndrieux reviewed (commented):
Overall LGTM

bert-e commented:
# Conflict




There is a conflict between your branch `improvement/ARTESCA-3572-handle-iso-removal` and the
destination branch `development/123.0`.

Please resolve the conflict on **the feature branch** (`improvement/ARTESCA-3572-handle-iso-removal`).

```sh
git fetch && \
git checkout origin/improvement/ARTESCA-3572-handle-iso-removal && \
git merge origin/development/123.0
```

Resolve merge conflicts and commit

```sh
git push origin HEAD:improvement/ARTESCA-3572-handle-iso-removal
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @alexandre-allard-scality

* @jbwatenbergscality

* @nicolast

* @gdemonet

* @teddyandrieux

* @thomasdanan

* @lucieleonard

* @chengyanjin

TeddyAndrieux commented on salt/metalk8s/archives/unmounted.sls:
```diff
@@ -0,0 +1,43 @@
+    - require:
+      - mount: Unmount '{{ archive.iso }}' from '{{ archive.path }}'
+
+Mountpoint '{{ archive.path }}' cannot be removed:
```
I would prefer a title that does not look like "an error" cause right now even if there is nothing in the directory in the salt output we will have "Mountpoint '{{ archive.path }}' cannot be removed" in the output.
(if `onlyif` is False the state is not executed but you still have it in the output with a comment saying "not executed because of onlyif")

I suggest 
```suggestion
Ensure '{{ archive.path }}' mountpoint is empty
```

gdemonet replied:
OK will do that. But to me, it's rather an issue with how the default output shows everything, even when skipped because of requisites.

gdemonet commented:
/bypass_author_approval

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/123.0`



The following branches will **NOT** be impacted:

* `development/2.0`
* `development/2.1`
* `development/2.10`
* `development/2.11`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`
* `development/2.6`
* `development/2.7`
* `development/2.8`
* `development/2.9`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **bypass_author_approval**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/123.0`



The following branches have **NOT** changed:

* `development/2.0`
* `development/2.1`
* `development/2.10`
* `development/2.11`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`
* `development/2.6`
* `development/2.7`
* `development/2.8`
* `development/2.9`



Please check the status of the associated issue ARTESCA-3572.

Goodbye sayf-eddine-scality.
